### ISSUE TITLE
Can't for the life of me get mini-profiler to run in production enviroment

### ISSUE THREAD

**Issue by Littlejd1997** [MiniProfiler/rack-mini-profiler#277]
Works fine in dev, and I have tried everything to get it to work in production
I have added this to an initalizer
```
Rack::MiniProfiler.config.enabled = true
Rack::MiniProfiler.config.pre_authorize_cb = lambda{|x| true}
```
As well as this as a before_action
```
def profiler
Rack::MiniProfiler.authorize_request
end
```
Here is the output of `rake middleware`
```
ubuntu@ip-10-0-0-48 ~/p/nuvola_academy> rake middleware
use Rack::MiniProfiler
use Rack::Cors
use Rack::Sendfile
use ActionDispatch::Static
use #<ActiveSupport::Cache::Strategy::LocalCache::Middleware:0x000000051db820>
use Rack::Runtime
use Rack::MethodOverride
use ActionDispatch::RequestId
use RequestStore::Middleware
use Rails::Rack::Logger
use ActionDispatch::ShowExceptions
use ActionDispatch::DebugExceptions
use ActionDispatch::RemoteIp
use ActionDispatch::Reloader
use ActionDispatch::Callbacks
use ActiveRecord::ConnectionAdapters::ConnectionManagement
use ActiveRecord::QueryCache
use ActionDispatch::Cookies
use ActionDispatch::Session::CookieStore
use ActionDispatch::Flash
use ActionDispatch::ParamsParser
use Rack::Head
use Rack::ConditionalGet
use Rack::ETag
use Warden::Manager
use Ahoy::Throttle
run NuvolaAcademy::Application.routes
```

I really have no idea what else to try

**Comment by matthewrk** [comment]
Try this in your initialiser:

```
require 'rack-mini-profiler'

Rack::MiniProfilerRails.initialize!(Rails.application)
```

And change your gem file entry to:

`gem 'rack-mini-profiler', require: false`

I came here because I was having the same issue, I had the above but needed some of the code you have to get it working :)

**Comment by Littlejd1997** [comment]
I must have forgotten to post that, but I already have both of those :(

**Comment by allthesignals** [comment]
Likewise can't get this to work in production. Anyone have anymore luck?

**Comment by SamSaffron** [comment]
This is concerning, can you repro this in a blank rails project? if so can
you send link to it so I can try it out?

**Comment by allthesignals** [comment]
@SamSaffron Hmmm, I figured out the project I think. I'm using [JSON API Resources gem](https://github.com/cerebris/jsonapi-resources), which inherits from a different controller base class. So those actions weren't being picked up by rack-mini-profiler. In addition to a regular HTML index page (that I had to create), I needed to add `Rack::MiniProfiler.authorize_request` to the API controllers as well (specifically, the `api_controller.rb`, the API equivalent of `application_controller.rb`).

**Comment by Littlejd1997** [comment]
It did work in a blank rails project, so theres gotta be something with this project. It's been around since rails 3, (now running 4.2.4) so it might have a few quirks. I have added the gemfile, as well as a link to the project live if that helps debugging. I still have no idea why it's not working.

Here is a link to the server (in production mode) that "should" have mini-rack-profiler running
http://54.157.238.172:443 (must use http and port 443)

```ruby '2.3.1'
source 'https://rubygems.org'
gem 'wicked_pdf'
gem 'wkhtmltopdf-binary'
gem 'puma'
gem 'timers'
gem 'pretender'
gem 'typhoeus'
gem 'by_star', :git => "https://github.com/radar/by_star"
gem 'iso_country_codes'
gem 'useragent'
gem 'tzinfo-data'
gem 'dotenv-rails'
gem 'rails', '~> 4.2.4'
gem 'active_model_serializers'
gem 'pg'
gem 'swagger-docs'
gem 'stripe'
gem 'ahoy_matey'
gem 'formtastic-bootstrap'
gem 'gravatar_image_tag'
gem "will_paginate", "~> 3.0.6"
gem 'will_paginate-bootstrap'
gem 'rack-cors', :require => 'rack/cors'
gem "annotate"
gem "yaml_db"
gem "paperclip", :git => "https://github.com/thoughtbot/paperclip.git"
gem 'aws-sdk'
gem 'rubyzip'
gem "jquery-fileupload-rails"
gem 'therubyracer', :platforms => :ruby
gem "carrierwave", :git => "https://github.com/carrierwaveuploader/carrierwave.git"
gem "fog"
gem 'nokogiri'
gem 'prawn'
gem 'httparty'
gem "font-awesome-rails"
gem "ransack"
gem 'newrelic_rpm'
gem "chartkick"
gem 'surus'

gem 'stackprof'     # For Ruby MRI 2.1+
gem 'rack-mini-profiler', require: false
gem 'memory_profiler'
gem 'flamegraph'

gem 'ckeditor'
gem 'carrierwave-aws'
gem 'dateslices'
gem 'turbolinks'
gem "cocoon"
gem 'jquery-minicolors-rails'
gem 'sitemap_generator'
gem 'mini_magick'
gem 'ims-lti',   "~> 1.0.2"
group :assets do
  gem 'sass-rails',   '~> 4.0.1'
  gem 'uglifier', '>= 1.0.3'
end

gem 'jquery-rails'
gem 'cancan'
gem 'devise', '>= 4.1.1'
gem 'rolify'
gem 'simple_form'

group :development, :test do
  gem 'better_errors'
  gem 'guard-bundler'
  gem 'guard-cucumber'
  gem 'guard-rails'
  gem 'guard-rspec'
  gem 'hub', :require=>nil
  #gem 'quiet_assets'
  gem 'rb-fchange', :require=>false
  gem 'rb-fsevent', :require=>false
  gem 'rb-inotify', :require=>false
  gem 'capybara'
  gem 'database_cleaner'
  gem 'email_spec'
  gem 'launchy'

  #gem 'binding_of_caller'
  gem 'factory_girl_rails'
  gem 'rspec-rails'
  gem 'debase'
  gem 'ruby-debug-ide'

  gem 'pry'
  gem 'pry-rails'
  gem 'rb-fsevent'

  #gem 'byebug'

  gem 'spring'
  gem 'swagger_client', :path => 'public/swagger/swagger_client'
end


gem 'nutus', '~> 2.0.0', :git => 'git@github.com:nuvolanetworks/nutus.git', :branch => 'master'
gem 'net-ssh'
gem 'sidekiq','~> 4.1'
gem 'streamio-ffmpeg'
gem 'sse-rails-engine'```

**Comment by ninjarabbi** [comment]
@Littlejd1997 Did you get it to work? I'm having trouble making it work in prod as well.

**Comment by rdlugosz** [comment]
I'm having a similar issue running in production. Works fine in Staging, which is also on Heroku. There are three main differences between our STG and PRD environments:
- Staging does not use SSL
- Staging does not use an asset host
- Staging does not use Honeybadger

I'm using the memory store & am doing the recommended stuff for injecting after Rack::Deflater, etc. 

Verified (via log statements) that the call to `authorize_request` is being made. Kind of running out of ideas; suspect it's somehow related to one of those three main differences listed above; will work through them and post here if I have any breakthroughs.

**Comment by Littlejd1997** [comment]
Hmm. I am not running on Heroku. My Rails server never does SSL (nginx does). Could be something with assets. I also don't use Honeybadger

**Comment by rdlugosz** [comment]
Well, it doesn't appear to be the asset host setting. I've enabled Cloudfront in our staging environment and RMP works fine in that environment.

**Comment by rdlugosz** [comment]
Ah! Sometimes all you need to do is take a walk and [re-read the Readme](https://github.com/MiniProfiler/rack-mini-profiler#storage)!

I fixed my issue by switching from `MemoryStore` to `RedisStore`. In my list above, I forgot one key difference between Staging and Production: _there are a lot of servers running in prod vs just one in staging_

Maybe this will fix your issue as well. Relevant section is here: https://github.com/MiniProfiler/rack-mini-profiler#storage

**Comment by SamSaffron** [comment]
I am closing this as stale, not  really sure what to do

**Comment by zachwalton** [comment]
This has been closed for a while, but for posterity since I just spent a few hours on this issue myself:

RackMiniProfiler stores cookies in whatever the defined backing store is, and skips profiling unless the user provides a valid cookie. In a distributed environment, this means that you'll never (or rarely) see profiling results as long as you're using the `LocalStorage` or `FileStorage` providers- not even for the current request.

IoW, if you want functional profiling in a distributed environment, use Redis or Memcached!
